### ISSUE TITLE
lm_to_projdata for the SAFIR listmode and with Cylindrical projdata template crashes

### ISSUE THREAD

**Issue by emikhay** [UCL/STIR#1077]
But everything works fine with the BlocksOnCylindrical projdata template.

**Comment by emikhay** [comment]
![stir_issue](https://user-images.githubusercontent.com/91138763/194287983-b91d6c61-0920-4980-a8b7-3d327224e24e.png)

**Comment by emikhay** [comment]
True! Yes, if I give it the crystal map everything works. Thank you so much, Kris

**Comment by KrisThielemans** [comment]
reopening as not really fixed in the code.

**Comment by KrisThielemans** [comment]
I didn’t write the SAFIR  code, but the main difficulty is that it serves 2 purposes:
- It allows reading a listmode file in SAFIR format
- It allows assigning LORs to a “virtual scanner”. (This was implemented before the “blocks” and “generic” scanners were made available).

My impression is that the “virtual scanner” functionality is enabled when there is a crystal map in the *SAFIR.par file: `CListEventSAFIR::get_bin()` will fall back to finding x,y,z from the crystal map, and then call `proj_data_info.get_bin()` (with proj_data_info initialised from the template), see [here](https://github.com/UCL/STIR/blob/7508babaf5e8147a3563f1642a47301af4c0834d/src/include/stir/listmode/CListRecordSAFIR.inl#L113-L116).

Checking the [alternative code](https://github.com/UCL/STIR/blob/7508babaf5e8147a3563f1642a47301af4c0834d/src/include/stir/listmode/CListRecordSAFIR.inl#L95-L109), it relies on [map_to_use()](https://github.com/UCL/STIR/blob/7508babaf5e8147a3563f1642a47301af4c0834d/src/include/stir/listmode/CListRecordSAFIR.h#L106). This is using the crystal map from the scanner. However, this is currently not set for cylindrical scanners #1078.

Current work-around is to use a crystal map in the *SAFIR.par file.

**Comment by chengqi001** [comment]
Using the formula “angle = math.atan2(y_scaled, x_scaled) % (2 * np.pi)”, the XYZ coordinates in list-mode can be mapped to ringA, ringB, detA, and detB, and then stored in the following format. This enables the conversion from list-mode to SAFIR-format, and subsequently to STIR sinograms.

https://github.com/UCL/STIR/blob/master/src/include/stir/listmode/CListRecordSAFIR.h
“
/*!
Provides interface of the record class to STIR by implementing get_LOR(). It uses an optional map from detector indices to
coordinates to specify LORAs2Points from given detection pair indices.

The record has the following format (for little-endian byte order)
\code
        unsigned ringA : 8;
        unsigned ringB : 8;
        unsigned detA : 16;
        unsigned detB : 16;
        unsigned layerA : 4;
        unsigned layerB : 4;
        unsigned reserved : 6;
        unsigned isDelayed : 1;
        unsigned type : 1;
\endcode
  \ingroup listmode
*/
”
For more details, refer to: [SAFIR Listmode Virtual Scanner Example](https://github.com/UCL/STIR/tree/master/examples/SAFIR-listmode-virtual-scanner).
